src/transactions.rs

```diff
@@ -0,0 +1,456 @@
+
+    /// Get a Transaction for a given TransactionHash
+    pub fn with_hash(&self, hash: &TransactionHash<L>) -> Result<Transaction<L>, KeystoreError<L>> {
+        let uid = self
+            .txn_by_hash
+            .get(hash)
+            .ok_or(KeyValueStoreError::KeyNotFound)?;
+        Ok(self.get(uid).unwrap())
+    }
+
+    /// Get a TransactionEditor for a given TransactionHash
+    pub fn with_hash_mut(
+        &mut self,
+        hash: &TransactionHash<L>,
+    ) -> Result<TransactionEditor<'_, L>, KeystoreError<L>> {
+        let txn = self.with_hash(hash)?;
+        Ok(TransactionEditor::new(&mut self.store, txn))
+    }
+
+    /// Get all the transactions timing out at the provided time.  There is no mutable way to get this
+    /// Collection because we should only be edditing one thing at a time
+    pub fn with_timeout(&self, timeout: u64) -> Result<Vec<Transaction<L>>, KeystoreError<L>> {
+        let uids = self
+            .expiring_txns
+            .get(&timeout)
+            .ok_or(KeyValueStoreError::KeyNotFound)?;
+
+        Ok(uids.iter().map(|uid| self.get(uid).unwrap()).collect())
+    }
+
+    /// Commit the store version.
+    pub fn commit(&mut self) -> Result<(), KeystoreError<L>> {
+        Ok(self.store.commit_version()?)
+    }
+
+    /// Revert the store version.
+    pub fn revert(&mut self) -> Result<(), KeystoreError<L>> {
+        Ok(self.store.revert_version()?)
+    }
+
+    /// Add a transaction has to the index and update the stored transaction with this has
+    pub fn insert_hash(
+        &mut self,
+        hash: TransactionHash<L>,
+        uid: &TransactionUID<L>,
+    ) -> Result<(), KeystoreError<L>> {
+        let editor = self.get_mut(uid)?;
+        editor.with_hash(hash.clone()).save()?;
+        self.txn_by_hash.insert(hash, uid.clone());
+        Ok(())
+    }
+
+    /// Add a list of Memo ids a transaction is waiting for.  We update the index and the stored transaction
+    pub fn add_pending_uids(
+        &mut self,
+        txn_uid: &TransactionUID<L>,
+        pending_uids: Vec<u64>,
+    ) -> Result<(), KeystoreError<L>> {
+        let txn_editor = self.get_mut(txn_uid)?;
+        txn_editor.add_pending_uids(&pending_uids).save()?;
+        for uid in pending_uids {
+            self.uids_awaiting_memos.insert(uid, txn_uid.clone());
+        }
+
+        Ok(())
+    }
+
+    /// Remove pending memo ids from the index and the stored transactions which were awaiting those ids
+    pub async fn remove_pending_uids(
+        &mut self,
+        pending_uids: Vec<u64>,
+    ) -> Result<(), KeystoreError<L>> {
+        for uid in pending_uids {
+            if let Some(txn_uid) = self.uids_awaiting_memos.remove(&uid) {
+                let txn_editor = self.get_mut(&txn_uid)?;
+                txn_editor.remove_pending_uid(uid).save()?;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Remove a transaction from the pending index when it
+    pub async fn remove_pending(
```

Comment:
I think it would be reasonable to have this just take a `timeout`, and remove _all_ the transactions with that timeout. Alternatively, we could have a virtual property `pending` where `TransactionEditor::set_pending(false)` just removes the transaction from the `expiring_txns` index.